File: reference/atip-discover/internal/validator/validator_test.go
package validator

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestNew(t *testing.T) {
	v, err := New()
	require.NoError(t, err)
	assert.NotNil(t, v)
}

func TestValidate_ValidMinimalMetadata(t *testing.T) {
	v, err := New()
	require.NoError(t, err)

	validJSON := `{
		"atip": {"version": "0.4"},
		"name": "test-tool",
		"version": "1.0.0",
		"description": "A test tool",
		"commands": {
			"run": {
				"description": "Run the tool",
				"effects": {
					"network": false
				}
			}
		}
	}`

	metadata, err := v.Validate([]byte(validJSON))
	require.NoError(t, err)
	assert.NotNil(t, metadata)
	assert.Equal(t, "test-tool", metadata.Name)
	assert.Equal(t, "1.0.0", metadata.Version)
	assert.Equal(t, "A test tool", metadata.Description)
}

func TestValidate_ValidComplexMetadata(t *testing.T) {
	v, err := New()
	require.NoError(t, err)

	validJSON := `{
		"atip": {"version": "0.4"},
		"name": "gh",
		"version": "2.45.0",
		"description": "GitHub CLI",
		"homepage": "https://cli.github.com",
		"trust": {
			"source": "native",
			"verified": true
		},
		"commands": {
			"pr": {
				"description": "Manage pull requests",
				"commands": {
					"list": {
						"description": "List pull requests",
						"options": [
							{
								"name": "state",
								"flags": ["-s", "--state"],
								"type": "enum",
								"enum": ["open", "closed", "merged"],
								"default": "open",
								"description": "Filter by state"
							}
						],
						"effects": {
							"network": true,
							"idempotent": true
						}
					}
				}
			}
		}
	}`

	metadata, err := v.Validate([]byte(validJSON))
	require.NoError(t, err)
	assert.NotNil(t, metadata)
	assert.Equal(t, "gh", metadata.Name)
}

func TestValidate_MissingRequiredFields(t *testing.T) {
	v, err := New()
	require.NoError(t, err)

	tests := []struct {
		name string
		json string
	}{
		{
			name: "missing atip field",
			json: `{"name": "tool", "version": "1.0.0", "description": "test"}`,
		},
		{
			name: "missing name field",
			json: `{"atip": {"version": "0.4"}, "version": "1.0.0", "description": "test"}`,
		},
		{
			name: "missing version field",
			json: `{"atip": {"version": "0.4"}, "name": "tool", "description": "test"}`,
		},
		{
			name: "missing description field",
			json: `{"atip": {"version": "0.4"}, "name": "tool", "version": "1.0.0"}`,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := v.Validate([]byte(tt.json))
			assert.Error(t, err)
		})
	}
}

func TestValidate_InvalidJSON(t *testing.T) {
	v, err := New()
	require.NoError(t, err)

	_, err = v.Validate([]byte("not valid json"))
	assert.Error(t, err)
}

func TestValidate_InvalidAtipVersion(t *testing.T) {
	v, err := New()
	require.NoError(t, err)

	invalidJSON := `{
		"atip": {"version": "99.99"},
		"name": "tool",
		"version": "1.0.0",
		"description": "test"
	}`

	_, err = v.Validate([]byte(invalidJSON))
	assert.Error(t, err)
}

func TestValidate_LegacyAtipFormat(t *testing.T) {
	v, err := New()
	require.NoError(t, err)

	// Legacy format: "atip": "0.3"
	legacyJSON := `{
		"atip": "0.3",
		"name": "tool",
		"version": "1.0.0",
		"description": "test",
		"commands": {
			"run": {
				"description": "Run",
				"effects": {"network": false}
			}
		}
	}`

	metadata, err := v.Validate([]byte(legacyJSON))
	require.NoError(t, err)
	assert.NotNil(t, metadata)
	assert.Equal(t, "tool", metadata.Name)
}

func TestValidate_InvalidEffects(t *testing.T) {
	v, err := New()
	require.NoError(t, err)

	tests := []struct {
		name string
		json string
	}{
		{
			name: "effects with invalid type",
			json: `{
				"atip": {"version": "0.4"},
				"name": "tool",
				"version": "1.0.0",
				"description": "test",
				"commands": {
					"run": {
						"description": "Run",
						"effects": {
							"destructive": "yes"
						}
					}
				}
			}`,
		},
		{
			name: "missing effects object",
			json: `{
				"atip": {"version": "0.4"},
				"name": "tool",
				"version": "1.0.0",
				"description": "test",
				"commands": {
					"run": {
						"description": "Run"
					}
				}
			}`,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := v.Validate([]byte(tt.json))
			assert.Error(t, err)
		})
	}
}

func TestValidate_PartialDiscovery(t *testing.T) {
	v, err := New()
	require.NoError(t, err)

	partialJSON := `{
		"atip": {"version": "0.4"},
		"name": "kubectl",
		"version": "1.28.0",
		"description": "Kubernetes CLI",
		"partial": true,
		"omitted": {
			"reason": "filtered",
			"safetyAssumption": "unknown"
		},
		"commands": {
			"get": {
				"description": "Get resources",
				"effects": {"network": true}
			}
		}
	}`

	metadata, err := v.Validate([]byte(partialJSON))
	require.NoError(t, err)
	assert.NotNil(t, metadata)
}

func TestParseJSON(t *testing.T) {
	validJSON := `{
		"atip": {"version": "0.4"},
		"name": "tool",
		"version": "1.0.0",
		"description": "test"
	}`

	metadata, err := ParseJSON([]byte(validJSON))
	require.NoError(t, err)
	assert.Equal(t, "tool", metadata.Name)
	assert.Equal(t, "1.0.0", metadata.Version)
}

func TestParseJSON_Invalid(t *testing.T) {
	_, err := ParseJSON([]byte("not json"))
	assert.Error(t, err)
}

func TestValidateMetadata(t *testing.T) {
	v, err := New()
	require.NoError(t, err)

	metadata := &AtipMetadata{
		Atip: map[string]interface{}{
			"version": "0.4",
		},
		Name:        "tool",
		Version:     "1.0.0",
		Description: "test",
		Commands: map[string]interface{}{
			"run": map[string]interface{}{
				"description": "Run",
				"effects": map[string]interface{}{
					"network": false,
				},
			},
		},
	}

	err = v.ValidateMetadata(metadata)
	assert.NoError(t, err)
}

func TestValidationError_Error(t *testing.T) {
	err := &ValidationError{
		Field:   "name",
		Message: "field is required",
	}

	assert.Contains(t, err.Error(), "name")
	assert.Contains(t, err.Error(), "required")
}

func TestNewWithSchema_CustomSchema(t *testing.T) {
	// Test with custom schema path
	v, err := NewWithSchema("/custom/schema.json")
	if err != nil {
		// Expected if file doesn't exist
		assert.Error(t, err)
	} else {
		assert.NotNil(t, v)
	}
}

func TestValidate_OptionsWithAllTypes(t *testing.T) {
	v, err := New()
	require.NoError(t, err)

	jsonWithOptions := `{
		"atip": {"version": "0.4"},
		"name": "tool",
		"version": "1.0.0",
		"description": "test",
		"commands": {
			"run": {
				"description": "Run",
				"options": [
					{
						"name": "string-opt",
						"flags": ["--str"],
						"type": "string",
						"description": "A string option"
					},
					{
						"name": "int-opt",
						"flags": ["--int"],
						"type": "integer",
						"description": "An integer option"
					},
					{
						"name": "bool-opt",
						"flags": ["--bool"],
						"type": "boolean",
						"description": "A boolean flag"
					},
					{
						"name": "enum-opt",
						"flags": ["--enum"],
						"type": "enum",
						"enum": ["a", "b", "c"],
						"description": "An enum option"
					}
				],
				"effects": {"network": false}
			}
		}
	}`

	metadata, err := v.Validate([]byte(jsonWithOptions))
	require.NoError(t, err)
	assert.NotNil(t, metadata)
}

func TestValidate_NestedCommands(t *testing.T) {
	v, err := New()
	require.NoError(t, err)

	nestedJSON := `{
		"atip": {"version": "0.4"},
		"name": "tool",
		"version": "1.0.0",
		"description": "test",
		"commands": {
			"level1": {
				"description": "Level 1",
				"commands": {
					"level2": {
						"description": "Level 2",
						"commands": {
							"level3": {
								"description": "Level 3",
								"effects": {"network": false}
							}
						}
					}
				}
			}
		}
	}`

	metadata, err := v.Validate([]byte(nestedJSON))
	require.NoError(t, err)
	assert.NotNil(t, metadata)
}

File: reference/atip-discover/tests/unit/validator.test.ts
import { describe, it, expect } from 'vitest';
import { validateMetadata } from '../../src/validator';

describe('ATIP Metadata Validation', () => {
  describe('validateMetadata', () => {
    it('should accept minimal valid ATIP metadata', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
        description: 'A test tool',
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(true);
      expect(result.errors).toEqual([]);
    });

    it('should accept legacy ATIP version format', () => {
      const metadata = {
        atip: '0.3',
        name: 'legacy-tool',
        version: '1.0.0',
        description: 'Legacy tool',
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(true);
    });

    it('should reject metadata without atip field', () => {
      const metadata = {
        name: 'test-tool',
        version: '1.0.0',
        description: 'Missing atip field',
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(false);
      expect(result.errors).toContainEqual(
        expect.objectContaining({
          path: ['atip'],
          message: expect.stringMatching(/required/i),
        })
      );
    });

    it('should reject metadata without name field', () => {
      const metadata = {
        atip: { version: '0.4' },
        version: '1.0.0',
        description: 'Missing name',
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(false);
      expect(result.errors).toContainEqual(
        expect.objectContaining({
          path: ['name'],
        })
      );
    });

    it('should reject metadata without version field', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        description: 'Missing version',
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(false);
      expect(result.errors).toContainEqual(
        expect.objectContaining({
          path: ['version'],
        })
      );
    });

    it('should reject metadata without description field', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(false);
      expect(result.errors).toContainEqual(
        expect.objectContaining({
          path: ['description'],
        })
      );
    });

    it('should accept metadata with commands', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
        description: 'Tool with commands',
        commands: {
          run: {
            description: 'Run command',
            effects: { network: false },
          },
        },
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(true);
    });

    it('should validate command effects', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
        description: 'Tool with effects',
        commands: {
          delete: {
            description: 'Delete data',
            effects: {
              destructive: true,
              reversible: false,
              network: false,
            },
          },
        },
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(true);
    });

    it('should reject commands with invalid effect types', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
        description: 'Invalid effects',
        commands: {
          run: {
            description: 'Run',
            effects: {
              destructive: 'maybe', // Should be boolean
            },
          },
        },
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(false);
    });

    it('should accept nested commands', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
        description: 'Tool with nested commands',
        commands: {
          pr: {
            description: 'Pull request commands',
            commands: {
              list: {
                description: 'List PRs',
                effects: { network: true },
              },
            },
          },
        },
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(true);
    });

    it('should validate option types', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
        description: 'Tool with options',
        commands: {
          run: {
            description: 'Run',
            options: [
              {
                name: 'output',
                flags: ['-o', '--output'],
                type: 'string',
                description: 'Output file',
              },
            ],
            effects: { network: false },
          },
        },
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(true);
    });

    it('should reject invalid option type', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
        description: 'Invalid option',
        commands: {
          run: {
            description: 'Run',
            options: [
              {
                name: 'bad',
                flags: ['--bad'],
                type: 'invalid-type', // Not a valid type
                description: 'Bad option',
              },
            ],
            effects: { network: false },
          },
        },
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(false);
    });

    it('should validate authentication metadata', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
        description: 'Tool with auth',
        authentication: {
          required: true,
          methods: [
            {
              type: 'oauth',
              setupCommand: 'test-tool auth login',
            },
          ],
          checkCommand: 'test-tool auth status',
        },
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(true);
    });

    it('should validate trust metadata', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
        description: 'Tool with trust info',
        trust: {
          source: 'native',
          verified: true,
        },
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(true);
    });

    it('should reject invalid trust source', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
        description: 'Invalid trust',
        trust: {
          source: 'unknown', // Should be 'native' or 'community'
          verified: true,
        },
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(false);
    });

    it('should validate partial discovery metadata', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test-tool',
        version: '1.0.0',
        description: 'Partial metadata',
        partial: true,
        filter: {
          commands: ['cmd1', 'cmd2'],
          depth: 2,
        },
        omitted: {
          reason: 'filtered',
          safetyAssumption: 'unknown',
        },
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(true);
    });

    it('should handle validation errors with multiple issues', () => {
      const metadata = {
        atip: { version: '0.4' },
        // Missing name, version, description
        commands: {
          bad: {
            // Missing description
            effects: {
              destructive: 'not-a-boolean', // Invalid type
            },
          },
        },
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(false);
      expect(result.errors.length).toBeGreaterThan(1);
    });

    it('should provide detailed error messages', () => {
      const metadata = {
        atip: { version: '0.4' },
        name: 'test',
        version: '1.0.0',
        // Missing description
      };

      const result = validateMetadata(metadata);

      expect(result.valid).toBe(false);
      expect(result.errors[0]).toHaveProperty('path');
      expect(result.errors[0]).toHaveProperty('message');
      expect(result.errors[0].message).toBeTruthy();
    });

    it('should reject non-object input', () => {
      const result = validateMetadata('not an object');

      expect(result.valid).toBe(false);
    });

    it('should reject null input', () => {
      const result = validateMetadata(null);

      expect(result.valid).toBe(false);
    });

    it('should reject array input', () => {
      const result = validateMetadata([]);

      expect(result.valid).toBe(false);
    });
  });
});
